File: ruby/get_user_info/get_user_info_better_version.rb
require 'rest-client'
require 'json'
require 'date'

class User
  REQUEST_TIMEOUT = 69

  # get user from the internet
  def fetch_users
    current_date = Time.now.strftime('%Y/%m/%d')
    # Come up with a better name for this variables.
    response_users = RestClient::Request.execute(
      method: :get, url: "http://example.com/users&starts_date=#{current_date}",
      timeout: REQUEST_TIMEOUT
    )
    JSON.parse response_users
  end

  def calculate_total_age
    users = get_users
    users.reduce(0){|total_age, user| total_age + calculate_age(user)}
  end

  def calculate_age user
    birthday = Date.strptime(user.fetch(:birth_day), '%Y')
    today = Date.today
    age = today.year - birthday.year
    age -= 1 if today < birthday + age.years
    age
  end
end

File: java/src/Part1/getuserinfo/GetUserInfoBetterVersion.java
package Part1.getuserinfo;

import java.util.Calendar;
import java.util.List;

public class GetUserInfoBetterVersion {

    private final int REQUEST_TIMEOUT = 69;

    static class User {
        String birthday;
    }

    static class RestClient {
        /**
         * DEMO METHOD
         * This method make a request to server then receive a list of users
         */
        static List<User> execute(String method, String url, int timeout) {
            return null;
        }
    }

    // get user from the internet
    private List<User> fetchUsers() {
        long currentTime = System.currentTimeMillis();
        return RestClient.execute("GET", "http://example.com/users&starts_date=" + currentTime, REQUEST_TIMEOUT);
    }

    private int calculateTotalAge() {
        List<User> users = fetchUsers();
        int total = 0;
        for (User user : users) {
            total += calculateAge(user);
        }
        return total;
    }

    private int calculateAge(User user) {
        int userBirthDay = getYearInTimeStr(user.birthday);
        int currentYear = Calendar.getInstance().get(Calendar.YEAR);
        return currentYear - userBirthDay;
    }

    /**
     * DEMO METHOD
     * This method will return year from time string
     */
    private static int getYearInTimeStr(String time) {
        return 0;
    }
}
